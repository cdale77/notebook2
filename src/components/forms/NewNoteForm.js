import React from "react";
import PropTypes from "prop-types";
import TextInput from "./TextInput";
import SubmitButton from "./SubmitButton";

class NewNoteForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = { name: "", expanded: false };
  }

  toggleExpander() {
    this.setState({ expanded: !this.state.expanded });
  }

  expanderClassName() {
    const base = "expander-trigger new-note-form-expander";
    if (this.state.expanded === true) return base;
    else return base + " expander-hidden";
  }

  handleNameChange(e) {
    this.setState({ name: e.target.value });
  }

  onSubmit(e) {
    e.preventDefault();
    const form = { name: this.state.name };
    this.props.onSubmit(form);
    this.setState({ expanded: false });
  }

  render() {
    return (
      <div className="expander">
        <button
          onClick={this.toggleExpander.bind(this)}
          className={"new-note-expander " + this.expanderClassName()}
        >
          + Add Note
        </button>
        <div className="new-note-form expander-content">
          <form>
            <div className="fieldset-wrapper">
              <TextInput
                name="name"
                title="Name"
                value={this.state.name}
                onChange={this.handleNameChange.bind(this)}
              />
              <SubmitButton onSubmit={this.onSubmit.bind(this)} />
            </div>
          </form>
        </div>
      </div>
    );
  }
}

NewNoteForm.propTypes = {
  onSubmit: PropTypes.func.isRequired
};

export default NewNoteForm;
