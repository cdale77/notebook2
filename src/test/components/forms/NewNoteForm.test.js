import React from "react";
import Renderer from "react-test-renderer";
import TestUtils from "react-dom/test-utils";
import NewNoteForm from "../../../components/forms/NewNoteForm";

describe("NewNoteForm", () => {
  describe("rendering", () => {
    it("should render the form", () => {
      const component = Renderer.create(<NewNoteForm onSubmit={jest.fn()} />);

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("submitting the form", () => {
    let callBack, spy, component;

    beforeEach(() => {
      callBack = {
        onSubmit: () => {}
      };

      spy = jest.spyOn(callBack, "onSubmit");
      component = TestUtils.renderIntoDocument(
        <NewNoteForm onSubmit={callBack.onSubmit} />
      );
    });

    it("should call the submit function", () => {
      const button = TestUtils.findRenderedDOMComponentWithClass(
        component,
        "button"
      );

      const nameField = TestUtils.findRenderedDOMComponentWithClass(
        component,
        "input-name"
      );

      // enter values in the fields to trigger the state change callbacks
      nameField.value = "changed";
      TestUtils.Simulate.change(nameField);

      TestUtils.Simulate.click(button);

      expect(spy).toHaveBeenCalled();
      expect(callBack.onSubmit.mock.calls.length).toEqual(1);

      const form = { name: "changed" };
      expect(callBack.onSubmit.mock.calls).toEqual([[form]]);
    });
  });
});
