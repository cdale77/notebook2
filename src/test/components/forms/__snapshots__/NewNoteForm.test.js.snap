// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewNoteForm rendering should render the form 1`] = `
<div
  className="expander"
>
  <button
    className="new-note-expander expander-trigger new-note-form-expander expander-hidden"
    onClick={[Function]}
  >
    + Add Note
  </button>
  <div
    className="form fieldset new-note-form expander-content"
  >
    <form>
      <div
        className="fieldset-wrapper"
      >
        <div
          className="form-group"
        >
          <input
            className="input-name"
            id="name"
            name="name"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="button"
            disabled={false}
            id="submit"
            name="submit"
            onClick={[Function]}
            type="submit"
            value="Submit"
          />
        </div>
      </div>
    </form>
  </div>
</div>
`;
